### PR TITLE
Move reflog into sqlite

### PR DESCRIPTION
**Pull request by ChrisPenner**
## Overview

Defines a format for keeping the reflog in sqlite;

* This ensures the reflog moves around with the codebase
* This guarantees that all references in the reflog refer to existing causals

(Ignore the weird duplications and such)

image

## Implementation notes

* Adds a new `reflog` table and an index on that table to sort by Date
* Adds a new migration which adds this table and ports an existing reflog to the db as best we can
* Rewrites the reflog output to use a table
* Note that the new reflog table only has (timestamp, new root causal hash, reason); we no longer track a `from` causal, since it will always just be the causal before it, and we don't show this in the output anyways.

When porting the old `reflog`:

* We throw out any malformed entries, or any entries which refer to causal hashes we don't have in the codebase
* We don't have timestamp info for old entries, so we just import them all at the "current time", but add second offsets such that they appear in the correct chronological order.
* I don't delete the old reflog file, I could, but there doesn't seem to be a good reason to and deleting files on the user's system is always a little bit sketch.

## Test coverage

Updated the reflog transcript

- - -
This change is [Reviewable](https://reviewable.io/reviews/unisonweb/unison/3364)

### CODE REVIEW

File: parser-typechecker/src/Unison/Codebase/Reflog.hs [hunk 1]
```diff
@@ -1,29 +0,0 @@
-{-# LANGUAGE ViewPatterns #-}
```
Comment (ChrisPenner) [inline parser-typechecker/src/Unison/Codebase/Reflog.hs:1]: Everywhere just uses the `Reflog.Entry` type from `U.Codebase.Reflog` now.

**Review by aryairani — CHANGES_REQUESTED** [review]
Awesome. A couple thoughts:

> Note that the new reflog table only has (timestamp, new root causal hash, reason); we no longer track a from causal, since it will always just be the causal before it, and we don't show this in the output anyways.

I think we still want to keep the old causal hash, just in case it somehow isn't the same as the entry before it. (The first entry has no entry before it, for example.)  Git is the same, in that it records both, but only shows the new hash in the output. We could also consider including both in the output (`#abc -> #def`), but it's messier from a "numbered output" perspective.

> We don't have timestamp info for old entries

Whoops!

> so we just import them all at the "current time", but add second offsets such that they appear in the correct chronological order.

Can there be a situation where I have 100 old entries and end up dating them into the future, and then a new user-generated entry is stuck in the middle of them (another case where I'd want my two causal hash ids per row again explicitly)? If it must be seconds and not nanoseconds or something (and even if it could be nanoseconds), putting them in the past would be safer.

> I don't delete the old reflog file, I could, but there doesn't seem to be a good reason to and deleting files on the user's system is always a little bit sketch.

:+1: If there's an easy spot to, you could tell the user we moved it and they can delete the file if they want.

**Comment by ChrisPenner** [comment]
> I think we still want to keep the old causal hash, just in case it somehow isn't the same as the entry before it

I don't have any issue with adding the `from`, but how could this happen and what does it mean? (besides the first entry of course)
If it just means we have a bug and forgot to add something to the reflog I'd rather we fix that error rather than try to cover it up 😓 

If we want confidence in that behaviour we could try adding a "reason" to the actual "updateRoot" command rather than having a separate "appendReflog" command.

> Can there be a situation where I have 100 old entries and end up dating them into the future

No, the second offsets go into the past.

**Comment by ChrisPenner** [comment]
@aryairani  

* I changed `putRootBranch` so it now accepts a reason and appends to the reflog in the same transaction which updates the root branch, so that should improve our confidence that the reflog is correct.
* I added the `from` causal to the reflog entries
* I added a note to the migration that the user can delete the old reflog file

~I just have to re-test things now 👍🏼~ ✅

**Comment by aryairani** [comment]
> I don't have any issue with adding the `from`, but how could this happen and what does it mean? (besides the first entry of course). If it just means we have a bug and forgot to add something to the reflog I'd rather we fix that error rather than try to cover it up 😓

Yeah.  It shouldn't happen (except in that case), and we should fix whatever errors we find. This won't be covering them up, but it will help us not lose data in those cases. It's not the strongest rationale, I know.  I wonder what the reasons for git also including the old and new hashes in their reflog were too, could be interesting.

File: parser-typechecker/src/Unison/Codebase/SqliteCodebase/Migrations/MigrateSchema5To6.hs [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+      pure entries
+  )
+    `catchIO` const (pure [])
+  where
+    parseEntry :: (Integer, Text) -> Maybe (Reflog.Entry CausalHash Text)
+    parseEntry (n, txt) =
+      -- We offset existing entries by a number of seconds corresponding to their position in
+      -- the current file; we can't reclaim timestamps for old reflog entries, but this at
+      -- least puts them in the correct order chronologically.
+      let offsetTime = addUTCTime (negate $ fromInteger @NominalDiffTime n) now
```
Comment (aryairani) [inline parser-typechecker/src/Unison/Codebase/SqliteCodebase/Migrations/MigrateSchema5To6.hs:54]: as a sanity check now, wdyt about adding a screenshot of a migrated reflog demoing this?

File: unison-src/transcripts/reflog.output.md [hunk 3]
```diff
@@ -66,10 +66,10 @@ y = 2
...
                               specified namespace.
   
-  1. #5fihla47kl : add
-  2. #2a4hnm3mcj : add
-  3. #43lqh9lfp3 : builtins.merge
-  4. #sg60bvjo91 : (initial reflogged namespace)
+       When   Action           Root Hash
+  1.   now    add              #5fihla47kl
+  2.   now    add              #2a4hnm3mcj
+  3.   now    builtins.merge   #43lqh9lfp3
```
Comment (aryairani) [inline unison-src/transcripts/reflog.output.md]: How hard would it be to restore the initial namespace to the output again?

File: parser-typechecker/src/Unison/Codebase/SqliteCodebase/Migrations/MigrateSchema5To6.hs (hunk 2 above)
Comment (aryairani) [reply parser-typechecker/src/Unison/Codebase/SqliteCodebase/Migrations/MigrateSchema5To6.hs:54]: ~~I guess the transcript does enough here, so you can disregard.~~

Comment (aryairani) [reply parser-typechecker/src/Unison/Codebase/SqliteCodebase/Migrations/MigrateSchema5To6.hs:54]: (Actually disregard the disregard; the transcript is unrelated to reflog migration)

**Comment by aryairani** [comment]
Note: I could imagine having some reason to do multiple `putRootBranch` in a single command that should possibly just get a single reflog update, but this seems good for now.

**Review by aryairani — CHANGES_REQUESTED** [review]
I left one request for display fix and one request for a sanity screenshot but otherwise looks good!

**Comment by ChrisPenner** [comment]
I verified that the entries from my personal reflog are being displayed in the correct order.

Here's the beginning and end of history;

image

image

**Comment by aryairani** [comment]
Sweet!

**Review by aryairani — APPROVED** [review]

**Review by aryairani — APPROVED** [review]

File: parser-typechecker/src/Unison/Codebase/SqliteCodebase/Migrations/MigrateSchema5To6.hs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+migrateCurrentReflog codebasePath = do
+  now <- Sqlite.unsafeIO $ getCurrentTime
+  oldEntries <- Sqlite.unsafeIO $ oldReflogEntries reflogPath now
+  for_ oldEntries \oldEntry -> do
+    -- There's no guarantee these causals actually exist in the DB,
+    -- so we check first to avoid triggering a bad foreign key constraint.
+    haveFrom <- isJust <$> Q.loadCausalByCausalHash (Reflog.fromRootCausalHash oldEntry)
+    haveTo <- isJust <$> Q.loadCausalByCausalHash (Reflog.toRootCausalHash oldEntry)
+    when (haveFrom && haveTo) $ Ops.appendReflog oldEntry
+  Sqlite.unsafeIO . putStrLn $ "I migrated old reflog entries from " <> reflogPath <> " into the codebase, you may delete that file now if you like."
```
Comment (aryairani) [inline parser-typechecker/src/Unison/Codebase/SqliteCodebase/Migrations/MigrateSchema5To6.hs]:
```suggestion
  Sqlite.unsafeIO . putStrLn $ "I migrated old reflog entries from " <> reflogPath <> " into the codebase; you may delete that file now if you like."
```

**Pull request merged**